File: React/src/features/users/data/datasources/UserRemoteDatasourceImpl.ts
// src/features/users/data/datasources/UserRemoteDatasourceImpl.ts

import { UserModel, userModelFromJson } from '../models/UserModel';
import { UserRemoteDatasource } from './UserRemoteDatasource';

export class UserRemoteDatasourceImpl implements UserRemoteDatasource {
  async getUsers(): Promise<UserModel[]> {
    const stopwatch = Date.now();

    console.log('========================');
    console.log('API REQUEST START');

    // Artificial delay — mirrors Flutter's Future.delayed(2s)
    await new Promise((resolve) => setTimeout(resolve, 2000));

    const response = await fetch(
      'https://jsonplaceholder.typicode.com/users',
    );

    const elapsed = Date.now() - stopwatch;

    console.log('========================');
    console.log('API REQUEST FINISHED');
    console.log(`TIME: ${elapsed} ms`);
    console.log('========================');

    const data = (await response.json()) as Record<string, unknown>[];

    return data.map((e) => userModelFromJson(e));
  }
}

File: Flutter/lib/features/users/data/datasources/user_remote_datasource_impl.dart
// lib/features/users/data/datasources/user_remote_datasource_impl.dart

import 'package:dio/dio.dart';

import '../models/user_model.dart';
import 'user_remote_datasource.dart';

class UserRemoteDataSourceImpl
    implements UserRemoteDataSource {

  final Dio dio;

  UserRemoteDataSourceImpl(this.dio);

  @override
  Future<List<UserModel>> getUsers() async {

    final stopwatch = Stopwatch()
      ..start();

    print(
      '========================',
    );

    print(
      'API REQUEST START',
    );

 

    final response = await dio.get(
      'https://jsonplaceholder.typicode.com/users',
    );

    stopwatch.stop();

    print(
      '========================',
    );

    print(
      'API REQUEST FINISHED',
    );

    print(
      'TIME: ${stopwatch.elapsedMilliseconds} ms',
    );

    print(
      '========================',
    );

final data = response.data as List;

final originalUsers = data.map(
  (e) => UserModel.fromJson(e),
).toList();

final users = <UserModel>[];

for (int i = 0; i < 1; i++) {

  for (final user in originalUsers) {

    users.add(
      UserModel(
        id: user.id + (i * 100),
        name: '${user.name} $i',
        username: user.username,
        email: user.email,
        phone: user.phone,
        website: user.website,
      ),
    );
  }
}

print(
  'TOTAL USERS GENERATED: ${users.length}',
);

return users;
  }
}